File: openai-agents/tour-of-agents/app/parallel_tools_agent.py
import restate

from agents import Agent, function_tool
from restate.ext.openai import restate_context, DurableRunner

from app.utils.utils import (
    InsuranceClaim,
    check_eligibility,
    compare_to_standard_rates,
    check_fraud,
)


# <start_here>
@function_tool
async def calculate_metrics(claim: InsuranceClaim) -> list[str]:
    """Calculate claim metrics."""
    ctx = restate_context()

    # Run tools/steps in parallel with durable execution
    results_done = await restate.gather(
        ctx.run_typed("eligibility", check_eligibility, claim=claim),
        ctx.run_typed("cost", compare_to_standard_rates, claim=claim),
        ctx.run_typed("fraud", check_fraud, claim=claim),
    )
    return [await result for result in results_done]


# <end_here>


agent = Agent(
    name="ParallelToolsAgent",
    instructions="You are a claim analysis agent that analyzes insurance claims."
    "Use the calculate_metrics tool and decide whether to approve.",
    tools=[calculate_metrics],
)


agent_service = restate.Service("ParallelToolClaimAgent")


@agent_service.handler()
async def run(_ctx: restate.Context, claim: InsuranceClaim) -> str:
    result = await DurableRunner.run(agent, f"Claim: {claim.model_dump_json()}")
    return result.final_output

File: vercel-ai/tour-of-agents/src/parallel-tools-agent.ts
import * as restate from "@restatedev/restate-sdk";
import { RestatePromise } from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import { generateText, tool, wrapLanguageModel, stepCountIs } from "ai";
import {
  ClaimInput,
  ClaimInputSchema,
  InsuranceClaim,
  InsuranceClaimSchema,
} from "./utils/types";
import {
  compareToStandardRates,
  checkEligibility,
  checkFraud,
} from "./utils/utils";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
const schema = restate.serde.schema;

// <start_here>
const run = async (ctx: restate.Context, claim: ClaimInput) => {
  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  const { text } = await generateText({
    model,
    prompt: `Analyze the claim ${JSON.stringify(claim)}.
        Use your tools to calculate key metrics and decide whether to approve.`,
    tools: {
      calculateMetrics: tool({
        description: "Calculate claim metrics.",
        inputSchema: InsuranceClaimSchema,
        execute: async (claim: InsuranceClaim) => {
          // Execute each calculation as a parallel durable step
          return RestatePromise.all([
            ctx.run("eligibility", () => checkEligibility(claim)),
            ctx.run("cost", () => compareToStandardRates(claim)),
            ctx.run("fraud", () => checkFraud(claim)),
          ]);
        },
      }),
    },
    stopWhen: [stepCountIs(10)],
    providerOptions: { openai: { parallelToolCalls: false } },
  });
  return text;
};
// <end_here>

const agent = restate.service({
  name: "ParallelToolClaimAgent",
  handlers: {
    run: restate.createServiceHandler({ input: schema(ClaimInputSchema) }, run),
  },
});

restate.serve({ services: [agent] });
